src/core/symbology/qgssymbollayerutils.h

```diff
@@ -400,25 +400,26 @@ class CORE_EXPORT QgsSymbolLayerUtils
      * \returns decoded symbol cast to specified type, if possible
      * \note not available in Python bindings
      */
-    template <class SymbolType> static SymbolType *loadSymbol( const QDomElement &element, const QgsReadWriteContext &context ) SIP_SKIP
+    template <class SymbolType> static std::unique_ptr< SymbolType > loadSymbol( const QDomElement &element, const QgsReadWriteContext &context ) SIP_SKIP
     {
-      QgsSymbol *tmpSymbol = QgsSymbolLayerUtils::loadSymbol( element, context );
-      SymbolType *symbolCastToType = dynamic_cast<SymbolType *>( tmpSymbol );
+      std::unique_ptr< QgsSymbol > tmpSymbol = QgsSymbolLayerUtils::loadSymbol( element, context );
+      const bool canCast = dynamic_cast<SymbolType *>( tmpSymbol.get() );
 
-      if ( symbolCastToType )
+      if ( canCast )
       {
-        return symbolCastToType;
+        std::unique_ptr< SymbolType > castRes( dynamic_cast<SymbolType *>( tmpSymbol.release() ) );
+        return castRes;
       }
       else
       {
         //could not cast
-        delete tmpSymbol;
         return nullptr;
       }
     }
 
     //! Reads and returns symbol layer from XML. Caller is responsible for deleting the returned object
-    static QgsSymbolLayer *loadSymbolLayer( QDomElement &element, const QgsReadWriteContext &context ) SIP_FACTORY;
```

Comment:
Why removing SIP_FACTORY ? How SIP would know that the ownership of the returned object is transferred to the callee?

Reply:
It's not needed -- the conversion code in conversions.sip knows there's ownership transferred when it's a unique ptr value